cmd/ocsp-updater/main.go

```diff
@@ -107,96 +214,102 @@ func (updater *OCSPUpdater) processResponse(tx *gorp.Transaction, serial string)
-		updater.stats.Inc("OCSP.Updates.Failed", 1, 1.0)
-		tx.Rollback()
-		return err
+		updater.log.AuditErr(err)
+		return
 	}
 
-	updater.log.Info(fmt.Sprintf("OCSP %s: OK", serial))
-	updater.stats.Inc("OCSP.Updates.Processed", 1, 1.0)
-	updater.stats.TimingDuration("OCSP.Updates.UpdateLatency", time.Since(innerStart), 1.0)
-	return nil
+	return
 }
 
-// findStaleResponses opens a transaction and processes up to responseLimit
-// responses in a single batch. The responseLimit should be relatively small,
-// so as to limit the chance of the transaction failing due to concurrent
-// updates.
-func (updater *OCSPUpdater) findStaleResponses(oldestLastUpdatedTime time.Time, responseLimit int) error {
-	var certificateStatus []core.CertificateStatus
-	_, err := updater.dbMap.Select(&certificateStatus,
-		`SELECT cs.* FROM certificateStatus AS cs JOIN certificates AS cert ON cs.serial = cert.serial
-		 WHERE cs.ocspLastUpdated < ? AND cert.expires > now()
-		 ORDER BY cs.ocspLastUpdated ASC
-		 LIMIT ?`, oldestLastUpdatedTime, responseLimit)
-
-	if err == sql.ErrNoRows {
-		updater.log.Info("All up to date. No OCSP responses needed.")
-	} else if err != nil {
-		updater.log.Err(fmt.Sprintf("Error loading certificate status: %s", err))
-	} else {
-		updater.log.Info(fmt.Sprintf("Processing OCSP Responses...\n"))
-		outerStart := time.Now()
-
-		for i, status := range certificateStatus {
-			updater.log.Debug(fmt.Sprintf("OCSP %s: (%d/%d)", status.Serial, i, responseLimit))
-
-			err = updater.updateOneSerial(status.Serial)
-			// Abort if we recieve a fatal error
-			if _, ok := err.(FatalError); ok {
-				return err
-			}
-		}
+// oldOCSPResponsesTick looks for certificates with stale OCSP responses and
+// generates/stores new ones
+func (updater *OCSPUpdater) oldOCSPResponsesTick(batchSize int, now, _ time.Time) {
+	updater.rpcMu.RLock()
 
-		updater.stats.TimingDuration("OCSP.Updates.BatchLatency", time.Since(outerStart), 1.0)
-		updater.stats.Inc("OCSP.Updates.BatchesProcessed", 1, 1.0)
+	statuses, err := updater.findStaleOCSPResponses(now.Add(-updater.ocspMinTimeToExpiry), batchSize)
+	if err != nil && err != errNoStaleResponses {
+		updater.stats.Inc("OCSP.Updates.Failed", 1, 1.0)
+		updater.log.AuditErr(err)
+		updater.rpcMu.RUnlock()
+		return
 	}
 
-	return err
+	updater.generateOCSPResponses(statuses)
+	updater.rpcMu.RUnlock()
 }
 
-func main() {
-	app := cmd.NewAppShell("ocsp-updater", "Generates and updates OCSP responses")
+type looper struct {
+	clk       clock.Clock
+	stats     statsd.Statter
+	batchSize int
+	tickDur   time.Duration
+	tickFunc  func(int, time.Time, time.Time)
+	name      string
+}
 
-	app.App.Flags = append(app.App.Flags, cli.IntFlag{
-		Name:   "limit",
-		Value:  100,
-		EnvVar: "OCSP_LIMIT",
-		Usage:  "Count of responses to process per run",
-	})
+func (l *looper) loop() {
+	prevTickStart := l.clk.Now().Add(-l.tickDur)
+	for {
```

Comment:
This will loop very, very quickly. Where is the sleep function use that will prevent these goroutines from eating up a core?